Advanced Dataframes Exercise Problems

In [128]:
import pandas as pd
import numpy as np
from env import host, user, password, get_db_url

Exercises Part 1

Problem 1 - 3

Create a function named get_db_url. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.

In [127]:
def get_db_url_local(username, hostname, password, db_name):
    url = f'mysql+pymysql://{username}:{password}@{hostname}/{db_name}'
    return url

This is in the env.py file.

Problem 4: Use your function to obtain a connection to the employees database.

In [5]:
url = get_db_url(user, host, password, 'employees')

In [6]:
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', url)

emp_no  birth_date first_name  last_name gender   hire_date
0   10051  1953-07-28   Hidefumi      Caine      M  1992-10-15
1   10052  1961-02-26     Heping     Nitsch      M  1988-05-21
2   10053  1954-09-13     Sanjiv   Zschoche      F  1986-02-04
3   10054  1957-04-04     Mayumi  Schueller      M  1995-03-13
4   10055  1956-06-06     Georgy     Dredge      M  1992-04-27

Problem 5:

a. Intentionally make a typo in the database url. What kind of error message do you see?

In [7]:
url = get_db_url(user, host, password, 'employs')

In [8]:
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', url)

OperationalError: (pymysql.err.OperationalError) (1044, "Access denied for user 'innis_1658'@'%' to database 'employs'")
(Background on this error at: https://sqlalche.me/e/14/e3q8)

Returns an access denied error

b. Intentionally make an error in your SQL query. What does the error message look like?

In [9]:
url = get_db_url(user, host, password, 'employees')

In [10]:
pd.read_sql('SELECT * FROM employes LIMIT 5 OFFSET 50', url)

ProgrammingError: (pymysql.err.ProgrammingError) (1146, "Table 'employees.employes' doesn't exist")
[SQL: SELECT * FROM employes LIMIT 5 OFFSET 50]
(Background on this error at: https://sqlalche.me/e/14/f405)

It access the database but since i mispelled the employees talbe in the query it lets me know the table does not exist.

Problem 6: Read the employees and titles tables into two separate DataFrames.

In [11]:
employees = pd.read_sql('SELECT * FROM employees', url)
employees

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

In [12]:
titles = pd.read_sql('SELECT * FROM titles', url)
titles

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

Problem 7: How many rows and columns do you have in each DataFrame? Is that what you expected?

In [13]:
employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300024 entries, 0 to 300023
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   emp_no      300024 non-null  int64 
 1   birth_date  300024 non-null  object
 2   first_name  300024 non-null  object
 3   last_name   300024 non-null  object
 4   gender      300024 non-null  object
 5   hire_date   300024 non-null  object
dtypes: int64(1), object(5)
memory usage: 13.7+ MB


I have 300024 rows and 6 columns in the employees DataFrame.

In [14]:
titles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB


I have 443308 rows and 4 columns in the titles DataFrame

Problem 8: Display the summary statistics for each DataFrame.

In [15]:
employees.describe()

emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000

In [16]:
titles.describe()

emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

Problem 9: How many unique titles are in the titles DataFrame?

In [17]:
len(pd.read_sql('SELECT title FROM titles GROUP BY title', url))

7

Problem 10: What is the oldest date in the to_date column?

In [18]:
pd.read_sql('SELECT to_date FROM titles', url).min()

to_date    1985-03-01
dtype: object

Problem 11: What is the most recent date in the to_date column?

In [19]:
pd.read_sql('SELECT to_date FROM titles WHERE to_date != "9999-01-01"', url).max()

to_date    2002-08-01
dtype: object

Exercise Part 2

Problem 1: Copy the users and roles DataFrames from the examples above.

In [20]:
# Create the users DataFrame.

users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users


id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [21]:
# Create the roles DataFrame

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles


id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

Problem 2: What is the result of using a right join on the DataFrames?

In [26]:
users.merge(roles, how='right',left_on='id',right_on='id')

id name_x  role_id     name_y
0   1    bob      1.0      admin
1   2    joe      2.0     author
2   3  sally      3.0   reviewer
3   4   adam      3.0  commenter

Problem 3: What is the result of using an outer join on the DataFrames?

In [27]:
users.merge(roles, how='outer',left_on='id',right_on='id')

id name_x  role_id     name_y
0   1    bob      1.0      admin
1   2    joe      2.0     author
2   3  sally      3.0   reviewer
3   4   adam      3.0  commenter
4   5   jane      NaN        NaN
5   6   mike      NaN        NaN

Problem 4: What happens if you drop the foreign keys from the DataFrames and try to merge them?

In [29]:
users.merge(roles, how='right')

id       name  role_id
0   1      admin      NaN
1   2     author      NaN
2   3   reviewer      NaN
3   4  commenter      NaN

Dropping both keys resutls in just the roles DataFrame with an extra NaN role_id column

In [30]:
users.merge(roles, how='right',left_on='id')

MergeError: Must pass "right_on" OR "right_index".

Dropping both results in an error. "Must Pass "right_on" OR "right_index"

Problem 5: Load the mpg dataset from PyDataset.

In [34]:
from pydataset import data
mpg = data('mpg')
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

Problem 6: Output and read the documentation for the mpg dataset.

In [36]:
data('mpg',show_doc = True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




Problem 7: How many rows and columns are in the dataset?

According to the documentation there are 234 rows and 11 columns

In [39]:
mpg.shape

(234, 11)

In [41]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


Shape and info() are other ways of getting more information on the DataFrame

Problem 8: Check out your column names and perform any cleanup you may want on them.

In [42]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


In [46]:
mpg = mpg.rename(columns={'displ':'engine displacement',
                          'cyl':'cylinders',
                          'trans':'transmission',
                          'cty':'city','hwy':'highway'})
mpg

manufacturer   model  engine displacement  year  cylinders transmission  \
1           audi      a4                  1.8  1999          4     auto(l5)   
2           audi      a4                  1.8  1999          4   manual(m5)   
3           audi      a4                  2.0  2008          4   manual(m6)   
4           audi      a4                  2.0  2008          4     auto(av)   
5           audi      a4                  2.8  1999          6     auto(l5)   
..           ...     ...                  ...   ...        ...          ...   
230   volkswagen  passat                  2.0  2008          4     auto(s6)   
231   volkswagen  passat                  2.0  2008          4   manual(m6)   
232   volkswagen  passat                  2.8  1999          6     auto(l5)   
233   volkswagen  passat                  2.8  1999          6   manual(m5)   
234   volkswagen  passat                  3.6  2008          6     auto(s6)   

    drv  city  highway fl    class  
1     f    18       29  p  compact  
2     f    21       29  p  compact  
3     f    20       31  p  compact  
4     f    21       30  p  compact  
5     f    16       26  p  compact  
..   ..   ...      ... ..      ...  
230   f    19       28  p  midsize  
231   f    21       29  p  midsize  
232   f    16       26  p  midsize  
233   f    18       26  p  midsize  
234   f    17       26  p  midsize  

[234 rows x 11 columns]

Problem 9: Display the summary statistics for the dataset.

In [47]:
mpg.describe()

engine displacement         year   cylinders        city     highway
count           234.000000   234.000000  234.000000  234.000000  234.000000
mean              3.471795  2003.500000    5.888889   16.858974   23.440171
std               1.291959     4.509646    1.611534    4.255946    5.954643
min               1.600000  1999.000000    4.000000    9.000000   12.000000
25%               2.400000  1999.000000    4.000000   14.000000   18.000000
50%               3.300000  2003.500000    6.000000   17.000000   24.000000
75%               4.600000  2008.000000    8.000000   19.000000   27.000000
max               7.000000  2008.000000    8.000000   35.000000   44.000000

Problem 10: How many different manufacturers are there?

In [227]:
len(mpg.groupby('manufacturer'))

15

Problem 11: How many different models are there?

In [83]:
len(mpg.groupby('model'))

38

Problem 12: Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.

In [85]:
mpg['milage_difference'] = mpg.highway - mpg.city
mpg

manufacturer   model  engine displacement  year  cylinders transmission  \
1           audi      a4                  1.8  1999          4     auto(l5)   
2           audi      a4                  1.8  1999          4   manual(m5)   
3           audi      a4                  2.0  2008          4   manual(m6)   
4           audi      a4                  2.0  2008          4     auto(av)   
5           audi      a4                  2.8  1999          6     auto(l5)   
..           ...     ...                  ...   ...        ...          ...   
230   volkswagen  passat                  2.0  2008          4     auto(s6)   
231   volkswagen  passat                  2.0  2008          4   manual(m6)   
232   volkswagen  passat                  2.8  1999          6     auto(l5)   
233   volkswagen  passat                  2.8  1999          6   manual(m5)   
234   volkswagen  passat                  3.6  2008          6     auto(s6)   

    drv  city  highway fl    class  milage_difference  
1     f    18       29  p  compact                 11  
2     f    21       29  p  compact                  8  
3     f    20       31  p  compact                 11  
4     f    21       30  p  compact                  9  
5     f    16       26  p  compact                 10  
..   ..   ...      ... ..      ...                ...  
230   f    19       28  p  midsize                  9  
231   f    21       29  p  midsize                  8  
232   f    16       26  p  midsize                 10  
233   f    18       26  p  midsize                  8  
234   f    17       26  p  midsize                  9  

[234 rows x 12 columns]

Problem 13: Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

In [109]:
mpg['average_mileage'] = mpg[['city','highway']].mean(axis=1)
mpg

manufacturer   model  engine displacement  year  cylinders transmission  \
1           audi      a4                  1.8  1999          4     auto(l5)   
2           audi      a4                  1.8  1999          4   manual(m5)   
3           audi      a4                  2.0  2008          4   manual(m6)   
4           audi      a4                  2.0  2008          4     auto(av)   
5           audi      a4                  2.8  1999          6     auto(l5)   
..           ...     ...                  ...   ...        ...          ...   
230   volkswagen  passat                  2.0  2008          4     auto(s6)   
231   volkswagen  passat                  2.0  2008          4   manual(m6)   
232   volkswagen  passat                  2.8  1999          6     auto(l5)   
233   volkswagen  passat                  2.8  1999          6   manual(m5)   
234   volkswagen  passat                  3.6  2008          6     auto(s6)   

    drv  city  highway fl    class  milage_difference  average_milage  \
1     f    18       29  p  compact                 11            23.5   
2     f    21       29  p  compact                  8            25.0   
3     f    20       31  p  compact                 11            25.5   
4     f    21       30  p  compact                  9            25.5   
5     f    16       26  p  compact                 10            21.0   
..   ..   ...      ... ..      ...                ...             ...   
230   f    19       28  p  midsize                  9            23.5   
231   f    21       29  p  midsize                  8            25.0   
232   f    16       26  p  midsize                 10            21.0   
233   f    18       26  p  midsize                  8            22.0   
234   f    17       26  p  midsize                  9            21.5   

     is_automatic  average_mileage  
1            True             23.5  
2           False             25.0  
3           False             25.5  
4            True             25.5  
5            True             21.0  
..            ...              ...  
230          True             23.5  
231         False             25.0  
232          True             21.0  
233         False             22.0  
234          True             21.5  

[234 rows x 15 columns]

Problem 14: Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission.

In [91]:
mpg['is_automatic'] = mpg.transmission.str.contains('auto')
mpg

manufacturer   model  engine displacement  year  cylinders transmission  \
1           audi      a4                  1.8  1999          4     auto(l5)   
2           audi      a4                  1.8  1999          4   manual(m5)   
3           audi      a4                  2.0  2008          4   manual(m6)   
4           audi      a4                  2.0  2008          4     auto(av)   
5           audi      a4                  2.8  1999          6     auto(l5)   
..           ...     ...                  ...   ...        ...          ...   
230   volkswagen  passat                  2.0  2008          4     auto(s6)   
231   volkswagen  passat                  2.0  2008          4   manual(m6)   
232   volkswagen  passat                  2.8  1999          6     auto(l5)   
233   volkswagen  passat                  2.8  1999          6   manual(m5)   
234   volkswagen  passat                  3.6  2008          6     auto(s6)   

    drv  city  highway fl    class  milage_difference  average_milage  \
1     f    18       29  p  compact                 11            23.5   
2     f    21       29  p  compact                  8            25.0   
3     f    20       31  p  compact                 11            25.5   
4     f    21       30  p  compact                  9            25.5   
5     f    16       26  p  compact                 10            21.0   
..   ..   ...      ... ..      ...                ...             ...   
230   f    19       28  p  midsize                  9            23.5   
231   f    21       29  p  midsize                  8            25.0   
232   f    16       26  p  midsize                 10            21.0   
233   f    18       26  p  midsize                  8            22.0   
234   f    17       26  p  midsize                  9            21.5   

     is_automatic  
1            True  
2           False  
3           False  
4            True  
5            True  
..            ...  
230          True  
231         False  
232          True  
233         False  
234          True  

[234 rows x 14 columns]

Problem 15: Using the mpg dataset, find out which which manufacturer has the best miles per gallon on average?

In [110]:
mpg.groupby('manufacturer').average_mileage.agg('mean').sort_values(ascending=False).head(1)


manufacturer
honda    28.5
Name: average_mileage, dtype: float64

Problem 16: Do automatic or manual cars have better miles per gallon?

In [111]:
mpg.groupby('is_automatic').average_mileage.agg('mean')

is_automatic
False    22.227273
True     19.130573
Name: average_mileage, dtype: float64

In [268]:
mpg.groupby('is_automatic')[['city','highway','average_mileage']].mean().sort_values(by= 'is_automatic',ascending=False)

city    highway  average_mileage
is_automatic                                       
True          15.968153  22.292994        19.130573
False         18.675325  25.779221        22.227273

In [269]:
mpg['Transmission_cat'] = np.where(mpg.transmission.str.contains('auto'),'auto','manual')

In [271]:
mpg.groupby('Transmission_cat').average_mileage.agg('mean')

Transmission_cat
auto      19.130573
manual    22.227273
Name: average_mileage, dtype: float64

Looks better

Manual cars have a better overall average mileage.

Exercises Part 3

Problem 1: Use your get_db_url function to help you explore the data from the chipotle database.

In [273]:
url = get_db_url('chipotle')
orders = pd.read_sql('SELECT * FROM orders',url,index_col='id')
orders

order_id  quantity                              item_name  \
id                                                                
1            1         1           Chips and Fresh Tomato Salsa   
2            1         1                                   Izze   
3            1         1                       Nantucket Nectar   
4            1         1  Chips and Tomatillo-Green Chili Salsa   
5            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4618      1833         1                          Steak Burrito   
4619      1833         1                          Steak Burrito   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   
4622      1834         1                     Chicken Salad Bowl   

                                                                     choice_description  \
id                                                                                        
1                                                                                   nan   
2                                                                          [Clementine]   
3                                                                               [Apple]   
4                                                                                   nan   
5            [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   
...                                                                                 ...   
4618  [Fresh Tomato Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce, Guacamole]]   
4619               [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese, Lettuce, Guacamole]]   
4620         [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]   
4621                                 [Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]   
4622                    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]   

     item_price  
id               
1        $2.39   
2        $3.39   
3        $3.39   
4        $2.39   
5       $16.98   
...         ...  
4618    $11.75   
4619    $11.75   
4620    $11.25   
4621     $8.75   
4622     $8.75   

[4622 rows x 5 columns]

In [220]:
orders.info()
orders

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4622 entries, 1 to 4622
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  4622 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 216.7+ KB


Problem 2: What is the total price for each order?

In [274]:
orders['item_price'] = orders['item_price'].str.replace('$','', regex =False).str.replace(',','', regex=False).astype('float')
orders.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4622 entries, 1 to 4622
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   order_id            4622 non-null   int64  
 1   quantity            4622 non-null   int64  
 2   item_name           4622 non-null   object 
 3   choice_description  4622 non-null   object 
 4   item_price          4622 non-null   float64
dtypes: float64(1), int64(2), object(2)
memory usage: 216.7+ KB


In [222]:
orders.groupby('order_id').item_price.agg('sum')

order_id
1       11.56
2       16.98
3       12.67
4       21.00
5       13.70
        ...  
1830    23.00
1831    12.90
1832    13.20
1833    23.50
1834    28.75
Name: item_price, Length: 1834, dtype: float64

Problem 3: What are the most popular 3 items?

In [228]:
orders.groupby('item_name').quantity.agg('sum').sort_values(ascending=False).head(3)

item_name
Chicken Bowl           761
Chicken Burrito        591
Chips and Guacamole    506
Name: quantity, dtype: int64

Problem 4: Which item has produced the most revenue?

In [229]:
orders.groupby('item_name').item_price.agg('sum').sort_values(ascending=False).head(1)

item_name
Chicken Bowl    7342.73
Name: item_price, dtype: float64

Problem 5: Join the employees and titles DataFrames together.

In [244]:
url = get_db_url('employees')
employees = pd.read_sql('DESCRIBE employees', url)
titles = pd.read_sql('Describe titles', url)
emp_titles = pd.read_sql('SELECT * FROM employees e INNER JOIN titles t ON e.emp_no = t.emp_no', url)

In [245]:
employees

Field           Type Null  Key Default Extra
0      emp_no            int   NO  PRI    None      
1  birth_date           date   NO         None      
2  first_name    varchar(14)   NO         None      
3   last_name    varchar(16)   NO         None      
4      gender  enum('M','F')   NO         None      
5   hire_date           date   NO         None

In [246]:
titles

Field         Type Null  Key Default Extra
0     emp_no          int   NO  PRI    None      
1      title  varchar(50)   NO  PRI    None      
2  from_date         date   NO  PRI    None      
3    to_date         date  YES         None

In [247]:
emp_titles

emp_no  birth_date first_name last_name gender   hire_date  emp_no  \
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26   10001   
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21   10002   
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28   10003   
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   10004   
4        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   10004   
...        ...         ...        ...       ...    ...         ...     ...   
443303  499997  1961-08-03    Berhard    Lenart      M  1986-04-21  499997   
443304  499997  1961-08-03    Berhard    Lenart      M  1986-04-21  499997   
443305  499998  1956-09-05   Patricia   Breugel      M  1993-10-13  499998   
443306  499998  1956-09-05   Patricia   Breugel      M  1993-10-13  499998   
443307  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30  499999   

                  title   from_date     to_date  
0       Senior Engineer  1986-06-26  9999-01-01  
1                 Staff  1996-08-03  9999-01-01  
2       Senior Engineer  1995-12-03  9999-01-01  
3              Engineer  1986-12-01  1995-12-01  
4       Senior Engineer  1995-12-01  9999-01-01  
...                 ...         ...         ...  
443303         Engineer  1987-08-30  1992-08-29  
443304  Senior Engineer  1992-08-29  9999-01-01  
443305     Senior Staff  1998-12-27  9999-01-01  
443306            Staff  1993-12-27  1998-12-27  
443307         Engineer  1997-11-30  9999-01-01  

[443308 rows x 10 columns]

I get the emp_no column twice so I will try a merge. 

In [248]:
url = get_db_url('employees')
employees = pd.read_sql('SELECT * FROM employees', url)
titles = pd.read_sql('SELECT * FROM titles', url)

In [249]:
employees

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

In [250]:
titles

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

In [253]:
emp_titles = employees.merge(titles, how='inner', on= 'emp_no')
emp_titles

emp_no  birth_date first_name last_name gender   hire_date  \
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26   
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21   
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28   
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
4        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
...        ...         ...        ...       ...    ...         ...   
443303  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443304  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443305  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443306  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443307  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30   

                  title   from_date     to_date  
0       Senior Engineer  1986-06-26  9999-01-01  
1                 Staff  1996-08-03  9999-01-01  
2       Senior Engineer  1995-12-03  9999-01-01  
3              Engineer  1986-12-01  1995-12-01  
4       Senior Engineer  1995-12-01  9999-01-01  
...                 ...         ...         ...  
443303         Engineer  1987-08-30  1992-08-29  
443304  Senior Engineer  1992-08-29  9999-01-01  
443305     Senior Staff  1998-12-27  9999-01-01  
443306            Staff  1993-12-27  1998-12-27  
443307         Engineer  1997-11-30  9999-01-01  

[443308 rows x 9 columns]

Problem 6: For each title, find the hire date of the employee that was hired most recently with that title.

In [254]:
emp_titles.groupby('title').hire_date.agg('max')

title
Assistant Engineer    1999-12-24
Engineer              2000-01-28
Manager               1992-02-05
Senior Engineer       2000-01-01
Senior Staff          2000-01-13
Staff                 2000-01-12
Technique Leader      1999-12-31
Name: hire_date, dtype: object

Problem 7: Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)

In [260]:
pd.read_sql('SHOW TABLES',url)

Tables_in_employees
0         departments
1            dept_emp
2        dept_manager
3           employees
4            salaries
5              titles

In [262]:
dep_titles = pd.read_sql('''SELECT * FROM departments d 
                                        Join dept_emp de 
                                            ON d.dept_no = de.dept_no 
                                        JOIN titles t 
                                            ON de.emp_no = t.emp_no''', url)
dep_titles                         

dept_no         dept_name  emp_no dept_no   from_date     to_date  \
0         d009  Customer Service   10011    d009  1990-01-22  1996-11-09   
1         d009  Customer Service   10038    d009  1989-09-20  9999-01-01   
2         d009  Customer Service   10038    d009  1989-09-20  9999-01-01   
3         d009  Customer Service   10049    d009  1992-05-04  9999-01-01   
4         d009  Customer Service   10049    d009  1992-05-04  9999-01-01   
...        ...               ...     ...     ...         ...         ...   
489898    d007             Sales  499986    d007  1985-08-11  9999-01-01   
489899    d007             Sales  499986    d007  1985-08-11  9999-01-01   
489900    d007             Sales  499987    d007  1999-12-21  9999-01-01   
489901    d007             Sales  499988    d007  1988-07-25  2001-10-09   
489902    d007             Sales  499988    d007  1988-07-25  2001-10-09   

        emp_no         title   from_date     to_date  
0        10011         Staff  1990-01-22  1996-11-09  
1        10038  Senior Staff  1996-09-20  9999-01-01  
2        10038         Staff  1989-09-20  1996-09-20  
3        10049  Senior Staff  2000-05-04  9999-01-01  
4        10049         Staff  1992-05-04  2000-05-04  
...        ...           ...         ...         ...  
489898  499986  Senior Staff  1992-08-11  9999-01-01  
489899  499986         Staff  1985-08-11  1992-08-11  
489900  499987         Staff  1999-12-21  9999-01-01  
489901  499988  Senior Staff  1997-07-25  2001-10-09  
489902  499988         Staff  1988-07-25  1997-07-25  

[489903 rows x 10 columns]

In [265]:
pd.crosstab(dep_titles.dept_name, dep_titles.title)

title               Assistant Engineer  Engineer  Manager  Senior Engineer  \
dept_name                                                                    
Customer Service                   298      2362        4             2027   
Development                       7769     58135        2            49326   
Finance                              0         0        2                0   
Human Resources                      0         0        2                0   
Marketing                            0         0        2                0   
Production                        6445     49649        4            42205   
Quality Management                1831     13852        4            11864   
Research                           378      2986        2             2570   
Sales                                0         0        2                0   

title               Senior Staff  Staff  Technique Leader  
dept_name                                                  
Customer Service           13925  16150               309  
Development                 1247   1424              7683  
Finance                    12139  13929                 0  
Human Resources            12274  14342                 0  
Marketing                  13940  16196                 0  
Production                  1270   1478              6557  
Quality Management             0      0              1795  
Research                   11637  13495               393  
Sales                      36191  41808                 0